lib/isexpired.c

```diff
@@ -69,10 +69,15 @@ int isexpired (const struct passwd *pw, /*@null@*/const struct spwd *sp)
 		return 1;
 	}
 
+	if (   __builtin_saddl_overflow(sp->sp_lstchg, sp->sp_max, &limit)
+	    || __builtin_saddl_overflow(limit, sp->sp_inact, &limit)) {
+		limit = LONG_MAX;
+	}
+
 	if (   (sp->sp_lstchg > 0)
 	    && (sp->sp_max >= 0)
 	    && (sp->sp_inact >= 0)
-	    && (now >= (sp->sp_lstchg + sp->sp_max + sp->sp_inact))) {
+	    && (now >= limit)) {
```

Comment:
Hm, I think I'm missing something here: you don't initialize limit to anything, so it is only set in the new stanza above if one of the additions overflows.  But I don't think you only check for (now >= limit) only in those cases.

Reply:
The result of __builtin_saddl_overflow is stored in its third argument, which is limit. So, limit is the result of adding sp->sp_lstchg, sp->sp_max and sp->sp_inact. If this calculation overflows, it is manually set to LONG_MAX.